tests/test_flow/fakes.py

```diff
@@ -67,132 +67,100 @@ def submit(self, task_config, fn, *args, **kwargs):
         ).submit()
 
 
-class FakeBlob:
-    def __init__(self, bucket, name, blobs):
-        assert not name.startswith("/")
-        self.bucket = bucket
-        self.name = name
-        self.blobs = blobs
-
-    def upload_from_string(self, data):
-        self.blobs[self._url] = data
-
-    def download_as_string(self):
-        return self.blobs[self._url]
-
-    def upload_from_filename(self, filename):
-        self.upload_from_string(open(filename, "r").read())
-
-    def download_to_filename(self, filename):
-        open(filename, "w").write(self.download_as_string())
-
-    def exists(self):
-        return self._url in self.blobs
-
-    def delete(self):
-        del self.blobs[self._url]
-
-    @property
-    def _url(self):
-        return f"gs://{self.bucket}/{self.name}"
-
-
-class FakeBucket:
-    def __init__(self, name, blobs):
-        self.name = name
-        self.blobs = blobs
-
-    def blob(self, blob_name):
-        return FakeBlob(self.name, blob_name, self.blobs)
-
-    def get_blob(self, blob_name):
-        return self.blob(blob_name)
-
-    def list_blobs(self, prefix):
-        assert not prefix.startswith("gs://")
-        return [
-            FakeBlob(self.name, k.replace(f"gs://{self.name}/", "", 1), self.blobs)
-            for (k, v) in self.blobs.items()
-            if k.startswith(f"gs://{self.name}/{prefix}")
-        ]
-
-
-class FakeGCS:
+class FakeGcsFs:
```

Comment:
Is there anything about this fake filesystem that's specific to GCS? Could we just use another FSSPEC implementation, like a `MemoryFileSystem` (maybe modified to not be global)?

Reply:
No, you're right. I should have thought of using `MemoryFileSystem` implementation! I think I implemented part of it in my PR 🤦. But what do you mean by not be global?

Reply:
From the `MemoryFileSystem` docs:

    This is a global filesystem so instances of this class all point to the same
    in memory filesystem.

We probably don't want this to be global if we're using it for testing. I suspect a subclass like this would work, but I could be wrong:

    class LocalMemoryFileSystem(MemoryFileSystem):
        def __init__(self):
            self.store = {}
            self.pseudo_dirs = []

Reply:
Actually, using `MemoryFileSystem` as a fake is getting annoying because things break with the gcs protocol `gs://` at the beginning of the url. I'll have to override all methods to remove / add protocol as required to make it work (I mean all and not only the ones that Bionic uses, since  `MemoryFileSystem` might be using some of those APIs underneath). There are some other small nuanced differences like for directories, `GcsFileSystem` return `True` whereas `MemoryFileSystem` returns false.

I think the current custom implementation is a better workaround than that.

Reply:
Due to nuanced differences, I don't think it would be easy to use fsspec like we initially thought either.